ffmpeg/ffmpeg.go

```diff
@@ -529,217 +763,26 @@ func (t *Transcoder) Transcode(input *TranscodeOptionsIn, ps []TranscodeOptions)
-				// & when B-frames are enabled nvenc struggles at calculating correct DTS
-				// XXX so we disable B-frames altogether to avoid PTS < DTS errors
-				if p.VideoEncoder.Opts["bf"] != "0" {
-					p.VideoEncoder.Opts["bf"] = "0"
-					glog.Warning("Forcing max_b_frames=0 for nvenc, as it can't handle those well with timestamp passthrough")
-				}
-			}
-		}
-		xcoderOutParams := C.CString(xcoderOutParamsStr)
-		defer C.free(unsafe.Pointer(xcoderOutParams))
-		gopMs := 0
-		if param.GOP != 0 {
-			if param.GOP <= GOPInvalid {
-				return nil, ErrTranscoderGOP
-			}
-			// Check for intra-only
-			if param.GOP == GOPIntraOnly {
-				p.VideoEncoder.Opts["g"] = "0"
-			} else {
-				if param.Framerate > 0 {
-					gop := param.GOP.Seconds()
-					interval := strconv.Itoa(int(gop * float64(param.Framerate)))
-					p.VideoEncoder.Opts["g"] = interval
-				} else {
-					gopMs = int(param.GOP.Milliseconds())
-				}
-			}
-		}
-		vidOpts := C.component_opts{
-			name: C.CString(encoder),
-			opts: newAVOpts(p.VideoEncoder.Opts),
-		}
-		audioEncoder := p.AudioEncoder.Name
-		if audioEncoder == "" {
-			audioEncoder = "aac"
-		}
-		audioOpts := C.component_opts{
-			name: C.CString(audioEncoder),
-			opts: newAVOpts(p.AudioEncoder.Opts),
-		}
-		fromMs := int(p.From.Milliseconds())
-		toMs := int(p.To.Milliseconds())
-		vfilt := C.CString(filters)
-		defer C.free(unsafe.Pointer(vidOpts.name))
-		defer C.free(unsafe.Pointer(audioOpts.name))
-		defer C.free(unsafe.Pointer(vfilt))
-		isDNN := C.int(0)
-		if p.Detector != nil {
-			isDNN = C.int(1)
-		}
-		params[i] = C.output_params{fname: oname, fps: fps,
-			w: C.int(w), h: C.int(h), bitrate: C.int(bitrate),
-			gop_time: C.int(gopMs), from: C.int(fromMs), to: C.int(toMs),
-			muxer: muxOpts, audio: audioOpts, video: vidOpts,
-			vfilters: vfilt, sfilters: nil, is_dnn: isDNN, xcoderParams: xcoderOutParams}
-		if p.CalcSign {
-			//signfilter string
-			escapedOname := ffmpegStrEscape(p.Oname)
-			signfilter := fmt.Sprintf("signature=filename='%s.bin'", escapedOname)
-			if p.Accel == Nvidia {
-				//hw frame -> cuda signature -> sign.bin
-				signfilter = fmt.Sprintf("signature_cuda=filename='%s.bin'", escapedOname)
-			}
-			sfilt := C.CString(signfilter)
-			params[i].sfilters = sfilt
-			defer C.free(unsafe.Pointer(sfilt))
-		}
-		defer func(param *C.output_params) {
-			// Work around the ownership rules:
-			// ffmpeg normally takes ownership of the following AVDictionary options
-			// However, if we don't pass these opts to ffmpeg, then we need to free
-			if param.muxer.opts != nil {
-				C.av_dict_free(&param.muxer.opts)
-			}
-			if param.audio.opts != nil {
-				C.av_dict_free(&param.audio.opts)
-			}
-			if param.video.opts != nil {
-				C.av_dict_free(&param.video.opts)
-			}
-		}(&params[i])
-	}
+
+    // Output configuration
+	params, finalizer, err := createCOutputParams(input, ps)
+    // This prevents C memory leaks
+	defer finalizer();
```

Comment:
There are tabs instead of spaces here. Also `;` at the end of line is not wanted.

You can run `gofmt -w ./ffmpeg/ffmpeg.go` manually to apply all those formatting rules go community enforces.

I got nice IDE to do this on every file save. Awesome tooling.

Reply:
Or just `go fmt ./...` in the project dir. We have it as a PR check in go-livepeer.